File: freecodecamp/2026-01-02.js
let dp = [];

function fooBar(n) {
  if (n <= 1) {
    return 0;
  }

  if (n == 2) {
    return 1;
  }

  if (dp[n] != null) {
    return dp[n];
  }

  return dp[n] = fooBar(n - 1) + fooBar(n - 2);
}

function nthFibonacci(n) {
  return fooBar(n);
}

File: freecodecamp/2026-01-02.py
dp = []

def foobar(n):
    if n <= 1:
        return 0
        
    if n == 2:
        return 1

    if dp[n] is not None:
        return dp[n]

    dp[n] = foobar(n - 1) + foobar(n - 2)
    return dp[n]

def nth_fibonacci(n):
    global dp
    dp = [None] * (n + 1)   # initialize memo table
    return foobar(n)
